src/main.cpp

```diff
@@ -638,22 +638,145 @@ unsigned int LimitOrphanTxSize(unsigned int nMaxOrphans) EXCLUSIVE_LOCKS_REQUIRE
     return nEvicted;
 }
 
-bool IsFinalTx(const CTransaction &tx, int nBlockHeight, int64_t nBlockTime)
+int64_t LockTime(const CTransaction &tx, int flags, const CCoinsView* pCoinsView, int nBlockHeight, int64_t nBlockTime)
 {
-    if (tx.nLockTime == 0)
-        return true;
-    if ((int64_t)tx.nLockTime < ((int64_t)tx.nLockTime < LOCKTIME_THRESHOLD ? (int64_t)nBlockHeight : nBlockTime))
-        return true;
-    BOOST_FOREACH(const CTxIn& txin, tx.vin)
-        if (!txin.IsFinal())
-            return false;
-    return true;
+    CCoins coins;
+
+    bool fEnforceBIP68 = static_cast<uint32_t>(tx.nVersion) >= 2
+                      && flags & LOCKTIME_VERIFY_SEQUENCE;
+
+    // Will be set to the equivalent height- and time-based nLockTime
+    // values that would be necessary to satisfy all relative lock-
+    // time constraints given our view of block chain history.
+    int nMinHeight = 0;
+    int64_t nMinTime = 0;
+    // Will remain equal to true if all inputs are finalized
+    // (CTxIn::SEQUENCE_FINAL).
+    bool fFinalized = true;
+
+    BOOST_FOREACH(const CTxIn& txin, tx.vin) {
+        // Set a flag if we witness an input that isn't finalized.
+        if (txin.nSequence == CTxIn::SEQUENCE_FINAL)
+            continue;
+        else
+            fFinalized = false;
+
+        // Do not enforce sequence numbers as a relative lock time
+        // unless we have been instructed to, and a view has been
+        // provided.
+        if (!(fEnforceBIP68 && pCoinsView))
+            continue;
+
+        // Sequence numbers equal to or above the locktime threshold
+        // are not treated as relative lock-times, nor are they given
+        // any consensus-enforced meaning at this point.
+        if (txin.nSequence >= CTxIn::SEQUENCE_LOCKTIME_THRESHOLD)
+            continue;
+
+        // Fetch the UTXO corresponding to this input.
+        if (!pCoinsView->GetCoins(txin.prevout.hash, coins)) {
+            // It is fully expected that coinbases inputs are not
+            // found in the UTXO set. Proceed to the next intput...
+            if (txin.prevout.IsNull())
+                continue;
+            // If a non-coinbase input cannot be found, we cannot
+            // be certain about whether lock-time constraints have
+            // been satisfied. Note that it should only ever be
+            // possible for this to happen with wallet transactions
+            // that have unknown inputs.
+            else
+                return std::numeric_limits<int64_t>::max();
```

Comment:
@maaku Did you test this in conjunction with a wallet? My guess is that it will start hiding all old transactions whose inputs are completely spent, because they are treated as non-final.